C++ interop: support for default arguments.

zygoloid opened the pull request:
The general strategy here is to force use of a thunk when we want to use default arguments, and have Clang generate uses of the default arguments on its side of the thunk.

To support this, change the key type used in `clang_decls` from being just a `Decl*` to being a pair of `Decl*` and number of parameters in the case of function decls. Import distinct `SemIR::Function`s for each number of parameters that's used, and corresponding distinct thunks.

zygoloid committed:
Support for default arguments.

The general strategy here is to force use of a thunk when we want to use
default arguments, and have Clang generate uses of the default arguments
on its side of the thunk.

To support this, change the key type used in `clang_decls` from being
just a `Decl*` to being a pair of `Decl*` and number of parameters in
the case of function decls. Import distinct `SemIR::Function`s for each
number of parameters that's used, and corresponding distinct thunks.

jonmeow commented on toolchain/sem_ir/clang_decl.h:
```diff
@@ -5,61 +5,92 @@
+
+// A Clang declaration mapped to a Carbon instruction.
+//
+// Instances of this type are managed by a `ClangDeclStore`, which ensures that
+// a single `ClangDecl` exists for each `ClangDeclKey` used.
+struct ClangDecl : public ClangDeclKey, public Printable<ClangDecl> {
```
Rather than making this inherit from `ClangDeclKey`, had you considered adding it as a member?

Notes on why I'm suggesting this:

- It's odd to see named initialization being removed, but I assume that's because named init doesn't really work well with inheritance (unless you do something like we did with `Function`, which seems excessive here).
- If you make it a member, I was thinking you get most of what you want, but the separation of what's decl-specific versus key-specific may be a little clearer (e.g. `PrintFields` could go away).
- Would make it clearer when you're switching types (because you'd have something  like `decl.key` as the parameter, instead of `decl`)
- Could this avoid having `operator==` and a hash if the key were fully separate?

zygoloid replied:
I think the main reason I prefer to use inheritance here is that I'm viewing the key/value distinction as being mostly secondary -- the primary thing we're doing is storing a set of `ClangDecl`s, which have a decl, an instruction, and maybe a parameter count, and we happen to provide indexing using a subset of those fields (usually the decl, but for functions also the parameter count). So I want all three fields to be direct members of `ClangDecl`. And I think I generally want an "isa" relationship between the two types.

Using a member would allow direct initialization `{.key = blah, .inst_id = blah}`, but not the kind that I'd want, which is `{.decl = blah, [.num_params = blah,] .inst_id = blah}`. I'd still want `PrintFields` so I can print a `ClangDecl` with three flattened fields without code duplication. And we would then need to provide a mixed-type `operator==` and a hasher for `ClangDecl` (those are indirectly required by `Carbon::Set`) rather than inheriting them from `ClangDeclKey`.

zygoloid replied:
Yeah, I went back and forth on this, and thinking about your comment I've gone back and forth a few more times :) I've ended up making this a member, though I feel conflicted about it! Happy to go back to the base class version if you prefer once you see this version.

I think the main reason I was preferring to use inheritance here is that I'm viewing the key/value distinction as being mostly secondary -- the primary thing we're doing is storing a set of `ClangDecl`s, which have a decl, an instruction, and maybe a parameter count, and we happen to provide indexing using a subset of those fields (usually the decl, but for functions also the parameter count). So I wanted all three fields to be accessible as direct members of `ClangDecl`; it makes me a little unhappy to access them with (eg) `.key.decl` rather than directly with `.decl`. But it's also a little sad that we have to duplicate the `==` and hashing logic if we make the key portion a member rather than a base class.

jonmeow replied:
Could add a `decl()` accessor, depending on your leanings -- still a little more syntax, but might make you a little happier not to see `.key`.

jonmeow commented on toolchain/sem_ir/clang_decl.h:
```diff
@@ -5,61 +5,92 @@
-  // Support direct comparison with the Clang AST node pointer.
-  auto operator==(const clang::Decl* rhs_decl) const -> bool {
-    return decl == rhs_decl;
+  auto PrintFields(llvm::raw_ostream& out) const -> void;
```
If you're keeping this (i.e. not making the key a member per other comment), please comment about what this does and why it's separate from `Print`

jonmeow commented on toolchain/sem_ir/clang_decl.h:
```diff
@@ -5,61 +5,92 @@
 
+  // The number of parameters to import for a function declaration. Always -1
+  // for a non-function declaration.
+  int params = -1;
```
Should we be explicit about size, since this is part of storage?

```suggestion
  int32_t params = -1;
```

Perhaps also `num_` to make it clearer at uses that it's a count, not a reference to actual params? (you also use `num_params` a few spots in this PR already, particularly `CalleeFunctionInfo`)

```suggestion
  int32_t num_params = -1;
```

zygoloid replied:
Done.

jonmeow commented on toolchain/sem_ir/clang_decl.h:
```diff
@@ -5,61 +5,92 @@
+  // count is expected.
+  template <typename DeclT>
+    requires(std::derived_from<DeclT, clang::Decl> &&
+             !std::derived_from<clang::FunctionDecl, DeclT> &&
+             !std::derived_from<DeclT, clang::FunctionDecl>)
```
Why does this check derived_from in both directions?

zygoloid replied:
We only want to use this constructor for declaration types that are unrelated to `FunctionDecl` (cannot possibly be a pointer to a `FunctionDecl`). A `BaseOfFunctionDecl*` could point to a `FunctionDecl`, and a `DerivedOfFunctionDecl*` definitely points to a `FunctionDecl`, so we need to check both to determine whether the type is unrelated.

jonmeow commented on toolchain/lower/file_context.cpp:
```diff
@@ -437,6 +437,20 @@ auto FileContext::BuildFunctionDecl(SemIR::FunctionId function_id,
     return nullptr;
   }
 
+  // Don't lower C++ function that use a thunk. We will never reference them
```
```suggestion
  // Don't lower C++ functions that use a thunk. We will never reference them
```

jonmeow commented on toolchain/sem_ir/clang_decl.h:
```diff
@@ -5,61 +5,92 @@
+  auto PrintFields(llvm::raw_ostream& out) const -> void;
+
+  auto operator==(const ClangDeclKey& rhs) const -> bool {
+    return decl == rhs.decl && params == rhs.params;
```
Perhaps use `= default;` here, instead of implementing?

zygoloid replied:
Unfortunately that doesn't work because the `Printable` base class doesn't provide an `==`. (Adding an `==` to `Printable` would be dangerous: it would make all `Printable`-derived classes comparable by default with an `==` that always returns `true`!)

jonmeow commented on toolchain/sem_ir/clang_decl.h:
```diff
@@ -5,61 +5,92 @@
+
+  // Factory function for clang declaration that is dynamically known to not be
+  // a function declaration.
+  static auto NonFunctionDecl(clang::Decl* decl) -> ClangDeclKey {
```
Take this comment with a little salt -- an extra action word would be useful for me and is hopefully at least neutral for you, but I'm not sure what you'll think of the API change and I don't intend to push for it.
 
Actin word: perhaps something like `Make...` or `For...` to include the action it's taking?

API change: I have some confusion reading calls to this, particularly right next to the non-function decl constructor on line 33. Just seeing call-sites, when there's `ClangDeclKey(...)` next to `ClangDeclKey::NonFunctionDecl(...)`, I'd typically assume that you're forming a function decl in the first case and a non-function decl in the second, but that's not the case.

IIUC, you're usually using this where you're already checking the decl kind. Maybe that check could be moved here? Perhaps something like:

```
  // Different API
  template<typename... DeclT>
  auto TryMakeNonFunctionDecl(Decl* decl) -> std::optional<ClangDeclKey> {
    if (!isa<DeclT...>(decl)) { return std::nullopt; }
    ...
  }

  // Example use
  if (auto key = TryMakeNonFunctionDecl<clang::NamespaceDecl, clang::TranslationUnitDecl>(parent_decl)) {
   ...
  }
```

zygoloid replied:
Renamed to `ForNonFunctionDecl`. I think that's consistent with what we've done in some other cases. I've also changed the `FunctionDecl` constructor into a factory function `ForFunctionDecl` for consistency and to make it a bit more explicit what's happening there.

One other thing that might make the call sites simpler for the reader would be to make the remaining constructor (from a non-function `Decl*`) be implicit, but I think the explicit construction here is not bad enough to deviate from the style rule :)

I think the logic in the caller is clearer if the check for decl kind is separate from the construction of the key, so I'm holding off on that change.

jonmeow commented on toolchain/check/cpp/import.cpp:
```diff
@@ -919,8 +920,8 @@ static auto ImportClassObjectRepr(Context& context, SemIR::ClassId class_id,
                          .index = SemIR::ElementIndex(fields.size())}));
-    context.clang_decls().Add(
-        {.decl = decl->getCanonicalDecl(), .inst_id = field_decl_id});
+    auto key = SemIR::ClangDeclKey::NonFunctionDecl(decl);
```
Does it work to use the `field` value here, or does that give a bad result for `IndirectFieldDecl`? i.e.:
```suggestion
    auto key = SemIR::ClangDeclKey(field);
```

zygoloid replied:
That doesn't work for indirect fields, added a comment to help future readers.

jonmeow commented on toolchain/check/cpp/import.cpp:
```diff
@@ -1727,16 +1732,20 @@ static auto AddDependentUnimportedTypeDecls(const Context& context,
+    const Context& context, const clang::FunctionDecl& clang_decl, int params,
     ImportWorklist& worklist) -> void {
-  for (const auto* param : clang_decl.parameters()) {
+  for (auto [i, param] : llvm::enumerate(clang_decl.parameters())) {
+    if (static_cast<int>(i) >= params) {
+      break;
+    }
```
Possible other approach:
```suggestion
  for (const auto* param : clang_decl.parameters().take_front(params)) {
```

zygoloid replied:
I don't really want to rely on `parameters()` returning an `ArrayRef` here -- I think that's more happenstance of the implementation than an API contract. I'd be happy to use a non-member `take_front` function, but LLVM doesn't seem to provide one (it does have `drop_begin` and `drop_end`, but using `drop_end` for this is awkward -- as well as being [inconsistently named](https://github.com/llvm/llvm-project/issues/160196)!).

What do you think about adding a `take_front` to common?

jonmeow replied:
I'd suggest writing it as `llvm::ArrayRef(clang_decl.parameters()).take_front` before a non-member function; if it changes from `ArrayRef`, maybe we can address it at that time?

zygoloid replied:
You want to see something awful? Here, this works:
```c++
  for (const auto* param :
       llvm::ArrayRef(clang_decl.parameters()).take_front(num_params)) {
```
but in thunk.cpp, the same thing does not work:
```c++
  for (const auto* callee_param : llvm::ArrayRef(callee_info.decl->parameters())
                                      .take_front(callee_info.num_params)) {
```
The reason is that in thunk.cpp, `callee_info.decl` is a pointer to a non-const `FunctionDecl`, so the return type of `parameters()` is a `MutableArrayRef` not an `ArrayRef`, and CTAD for `llvm::ArrayRef` deduces `llvm::ArrayRef<llvm::MutableArrayRef<ParmVarDecl*>>` , using the `ArrayRef` constructor that makes a single-element `ArrayRef` from a single value.

Anyway... the nearby TODOs point out that we don't actually want to be iterating over `parameters` at all; I've fixed the TODOs by grabbing the semantic parameter types instead of the as-declared types, and switched to `seq` :) Newly-added test verifies we're now properly handling parameters with qualified and decayed types.

jonmeow replied:
The getParamType changes moot this, but a few things:

1) I'm mainly responding for your "awful" comment. I think `makeArrayRef` works the way you want.

2) I meant to just write `parameters().take_front` right now, without a constructor call. I wasn't advocating adding a constructor call if not needed. I'm more on the end of assuming the API is stable, and changes (and what fixes would be needed) are hard to predict; but perhaps you have information I don't, about plans to change it?

zygoloid replied:
FWIW, `makeArrayRef` doesn't exist any more; it was removed in favor of CTAD.

jonmeow commented on toolchain/check/cpp/import.cpp:
```diff
@@ -1727,16 +1732,20 @@ static auto AddDependentUnimportedTypeDecls(const Context& context,
 // and adds them to the given set.
 static auto AddDependentUnimportedFunctionDecls(
-    const Context& context, const clang::FunctionDecl& clang_decl,
+    const Context& context, const clang::FunctionDecl& clang_decl, int params,
```
Perhaps pass as `num_params` to distinguish more from `parameters`?

```suggestion
    const Context& context, const clang::FunctionDecl& clang_decl, int num_params,
```

jonmeow commented on toolchain/check/cpp/thunk.cpp:
```diff
@@ -243,8 +262,11 @@ static auto BuildThunkParameterTypes(clang::ASTContext& ast_context,
 
-  for (const clang::ParmVarDecl* callee_param :
-       callee_info.decl->parameters()) {
+  for (auto [i, callee_param] :
+       llvm::enumerate(callee_info.decl->parameters())) {
+    if (static_cast<int>(i) >= callee_info.num_params) {
+      break;
+    }
```
Just suggesting the different approach, copying from the earlier state
```suggestion
  for (const clang::ParmVarDecl* callee_param :
       callee_info.decl->parameters().take_front(callee_info.num_params)) {
```

jonmeow commented on toolchain/check/cpp/thunk.cpp:
```diff
@@ -415,8 +438,8 @@ static auto BuildCalleeArgs(clang::Sema& sema,
-  unsigned first_param = callee_info.has_explicit_object_parameter();
-  unsigned num_params = callee_info.decl->getNumParams();
+  int first_param = callee_info.has_explicit_object_parameter();
+  int num_params = callee_info.num_params;
```
You might just remove this local (num_params), I assume it was added because the previous call was longer (and a function call)

jonmeow commented on toolchain/check/testdata/interop/cpp/function/default_arg.carbon:
```diff
@@ -0,0 +1,618 @@
+  // //@dump-sem-ir-begin
+  Cpp.A(1, 2);
+  Cpp.A(1, 2, 3, 4);
+  // Cpp.A(1, 2, 3);
+  // ({} as Cpp.X).B(1);
+  // Cpp.X.C(1);
+  // ({} as Cpp.X).D(1);
```
Why is this commented out?

zygoloid replied:
Oops, I commented this out while debugging and forgot to uncomment. Fixed, thanks!

jonmeow commented on toolchain/check/testdata/interop/cpp/class/constructor.carbon:
```diff
@@ -186,11 +181,6 @@ import Cpp library "implicit_multi_arguments.h";
 fn F() {
   //@dump-sem-ir-begin
   let c1: Cpp.C = Cpp.C.C(8, 9);
-  // CHECK:STDERR: fail_todo_import_implicit_multi_arguments.carbon:[[@LINE+5]]:19: error: 1 argument passed to function expecting 2 arguments [CallArgCountMismatch]
```
I assume it's expected this constructor test still doesn't work?

zygoloid replied:
Yes, we don't use C++ constructors for Carbon conversions yet.

jonmeow reviewed (commented):
Nice job getting this working!

zygoloid committed:
Adjust name and type of `ClangDeclKey::params`.

zygoloid committed:
Rename `NonFunctionDecl` -> `ForNonFunctionDecl`.

zygoloid committed:
Apply suggestions from code review.

zygoloid committed:
Switch key from base to member.

Print clang decls when printing SemIR file and add test.

zygoloid committed:
Comments added and removed here.

zygoloid committed:
Access parameter types by index.

In passing, this fixes a failure to properly decay parameter types when
importing them.

zygoloid committed:
Add more testing of parameter decay.

Fix diagnostics being emitted in the wrong order, exposed by testing.

zygoloid committed:
clang-format

zygoloid committed:
Rename ClangDeclKey constructor for functions to ClangDeclKey::ForFunctionDecl.

jonmeow commented on toolchain/check/cpp/import.cpp:
```diff
@@ -1342,21 +1347,28 @@ static auto MakeImplicitParamPatternsBlockId(
-  if (clang_decl.parameters().empty()) {
+                                     const clang::FunctionDecl& clang_decl,
+                                     int num_params) -> SemIR::InstBlockId {
+  if (clang_decl.parameters().empty() || !num_params) {
```
Minor, but I don't think as much looking at an integer comparison as looking at an implicit cast of an integer to a bool.

```suggestion
  if (clang_decl.parameters().empty() || num_params == 0) {
```

jonmeow commented on toolchain/check/cpp/import.cpp:
```diff
@@ -1634,34 +1647,36 @@ static auto ImportFunction(Context& context, SemIR::LocId loc_id,
 
-auto ImportCppFunctionDecl(Context& context, SemIR::LocId loc_id,
-                           clang::FunctionDecl* clang_decl) -> SemIR::InstId {
+static auto ImportFunctionDecl(Context& context, SemIR::LocId loc_id,
```
Since you're creating this function, would it be worth some comments on behavior?

jonmeow reviewed (approved):
LG

zygoloid committed:
Explicitly compare against zero.

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

zygoloid committed:
Add comment.

zygoloid committed:
Merge branch 'trunk' into toolchain-default-arguments

zygoloid committed:
Fix unit test

zygoloid committed:
Merge branch 'trunk' into toolchain-default-arguments
